### ISSUE TITLE
conversion from command line hangs on linux

### ISSUE THREAD

**Issue by chubinou** [umlet/umlet#320]
umlet  hangs while converting uxf files from the command line.

> umlet -action=convert -format=png -filename=test.uxf -output=test.png
> Conversion finished: "/home/user/./test.uxf" to "test.png"
> [... hangs here ...]

the file is test.png is created successfully though

tested on debian (testing), umlet version 13.3 stand alone.
> java -version 
> java version "1.7.0_85"
> OpenJDK Runtime Environment (IcedTea 2.6.1) (7u85-2.6.1-3)
> OpenJDK 64-Bit Server VM (build 24.85-b03, mixed mode)

this might be related to #310 and #307

**Comment by afdia** [comment]
Hi
I have tried:
sh umlet.sh -action=convert -format=png -filename=palettes/Plots.uxf -output=test.png
java -jar umlet.jar -action=convert -format=png -filename=palettes/Plots.uxf -output=test.png
and both commands convert and terminate correctly on Ubuntu 12.04 with java 1.7.0_75

Do you use the shell script or the umlet.jar directly?

**Comment by chubinou** [comment]
I'm using this version:
https://packages.debian.org/stretch/umlet
which is the version 13.3 of umlet

umlet is started by calling the jar directly 

umlet version in ubuntu 12.04 repository is 11.3. Did you test this version or the 13.3?

Though, we've test with archlinux version (also 13.3), and it ran successfully, so it might be a debian bug

**Comment by smumm** [comment]
Hi,

tested it under Arch and it was working without any issues. Thought me prior debian testing installation was also working. Maybe the current debian package on testing is not correct but it should be the same as debian unstable (used by Ubuntu).

Can you download version 13.3 from www.umlet.com and start it via the script umlet.sh?

**Comment by chubinou** [comment]
I have the same behavior (the conversion is done but the program hangs). I also tried to remove my ~/.umlet folder.

is there some logs/backtrace I can provide?

**Comment by afdia** [comment]
Do you mean that the error also occurs if you use UMLet 13.3 from www.umlet.com?
We do not distribute UMLet on other sites, therefore the debian package seems to be built by other guys and could potentially behave differently.

**Comment by chubinou** [comment]
yes, I have the same issue with the binary downloaded from www.umlet.com

**Comment by chubinou** [comment]
I ran the test with older versions of umlet (13.2,  12.2, 11.0) and have the same behavior. It used to works with theses versions. So I guess the problem might not be on your side.

**Comment by afdia** [comment]
OK, I guess I can close the issue until we find some problem which can be solved in umlets code

**Comment by JensErat** [comment]
I experience the exactly same issue on Debian testing, running umlet from either the download web page or the Debian repository as linked above.

Conversion actually takes place, the output file is created (and contains the diagram), but umlet never quits. Is there any command line flag to provide a debug trace or similar help at reproducing/finding the issue?

Jave version in use:

    $ java -version
    java version "1.7.0_85"
    OpenJDK Runtime Environment (IcedTea 2.6.1) (7u85-2.6.1-3)
    OpenJDK 64-Bit Server VM (build 24.85-b03, mixed mode)

I use Kernel 4.2 from the Debian repositories, but don't think that this might be linked with the issue:

    $ uname -a
    Linux t440s 4.2.0-trunk-amd64 #1 SMP Debian 4.2-1~exp1 (2015-08-31) x86_64 GNU/Linux

**Comment by afdia** [comment]
thanks for you report, I've reopened the issue to see if it can be fixed on out side.

There is no such flag, but there are 3 external visible steps in the command line call:
1. the outputfile will be generated
2. you should see a message like "Conversion finished: "D:\UML Class.uxf" to "D:\output.jpg""
3. program terminates

(This sample is based on the following call: "-action=convert -format=jpg "-filename=D:/UML Class.uxf" -output=D:\output.jpg")

The strange thing is that this has not been reported before, so perhaps it is influenced by the Java version or the OS.
If you have the time and some programming experience, you could also try to check out the source code and start the batchmode in debugmode within Eclipse to see if and where the conversion hangs

**Comment by JensErat** [comment]
After configuring eclipse for UMLet (and realizing I had to install the PDE), the problem disappeared.

I deleted the PDE, so there is no hidden dependency; but I realized I had some openjdk-update between reporting the issue and using it.

@chubinou Does the issue still affect you after running a (dist-)upgrade?

**Comment by chubinou** [comment]
The problem disappeared also on my side.

**Comment by afdia** [comment]
I will leave the issue open for some time but it seems that it's really not under our control

thanks for your help in analyzing the issue!

**Comment by JensErat** [comment]
Today I ran into the issue again. This time, I was even able to reproduce it in debug mode. After conversion, three threads are still running:

- AWT-EventQueue-0
- Thread-0
- DestroyJavaVM

Suspending latter two does not result in a stack trace, these seem to be rather terminated. Looking at the AWT thread, it seems that AWT is stuck in a deadlock:

    Thread [AWT-EventQueue-0] (Suspended)	
    	Unsafe.park(boolean, long) line: not available [native method]	
    	LockSupport.park(Object) line: 186	
    	AbstractQueuedSynchronizer$ConditionObject.await() line: 2043	
    	EventQueue.getNextEvent() line: 555	
    	EventDispatchThread.pumpOneEventForFilters(int) line: 211	
    	EventDispatchThread.pumpEventsForFilter(int, Conditional, EventFilter) line: 161	
    	EventDispatchThread.pumpEventsForHierarchy(int, Conditional, Component) line: 150	
    	EventDispatchThread.pumpEvents(int, Conditional) line: 146	
    	EventDispatchThread.pumpEvents(Conditional) line: 138	
    	EventDispatchThread.run() line: 91	

Software releases I'm currently running (`i3` is my "window" manager, which is a tiling window manager -- sometimes these result in funny effects when software doesn't handle edge cases correctly).

    $ java -version
    java version "1.7.0_91"
    OpenJDK Runtime Environment (IcedTea 2.6.3) (7u91-2.6.3-1)
    OpenJDK 64-Bit Server VM (build 24.91-b01, mixed mode)
    
    $ uname -a
    Linux t440s 4.2.0-1-amd64 #1 SMP Debian 4.2.6-1 (2015-11-10) x86_64 GNU/Linux
    
    $ i3 --version
    i3 version 4.11 (2015-09-30, branch "4.11") © 2009 Michael Stapelberg and contributors

**Comment by JensErat** [comment]
Looking further through some Google results, it seems umlet might not correctly initialize some UI objects.

**Comment by afdia** [comment]
Are you able to reproduce the issue in a certain way or does it happen sporadically without a pattern?

Could you elaborate what you mean with "umlet might not correctly initialize some UI objects"
Unfortunately it's not always easy to find the correct way to do things in Swing ;)

**Comment by JensErat** [comment]
It just happens in my machine, and it happens always.

I haven't really analyzed the issue. But from what I see (and a short Google query with some keywords I don't know any more, I think it was about the thread still active) it looks like some Swing race condition occurring when GUI elements are not correctly set up or torn down. Sadly debugging the issue isn't easy -- the dependencies on the Eclipse development framework break other projects for me, and debugging the issue would require digging deep into frameworks I never worked with.

As an educated guess I would say some state or action when the GUI framework is instanciated/torn down is not observed. "Normal" window managers might handle this (or do some other surprising magic, have different defaults for some values, ...), but the rather limited one I'm using differs here. Sadly I haven't observed issues when using the graphical user interface and no idea what exactly is going wrong here.

> Unfortunately it's not always easy to find the correct way to do things in Swing ;)

Anyway: This also seems an architectural issue. When working on the command line, is it really required to load all those GUI stuff? I'd guess the right way to do things in Swing for command line usage would not doing it at all. :)  Furthermore, is it really necessary to have Eclipse dependencies in the core code, couldn't this be somehow extracted to facilitate development on "other parts" of the code? I don't know how easy it would be to pull out those dependencies, but I'd guess it might make both development easier and reduce the number of potential issues if dependencies not required for some parts are moved out. If the Eclipse dependencies are just about integration of some panels in Eclipse and maybe some file handling, I'd guess it should be rather easy.

Loading both GUI components in general and especially the Eclipse framework involves lots of magic happening behind the scenes hard to understand, and possibly this issue is a great example what can go wrong. :)

**Comment by afdia** [comment]
Hi

In the last weeks we have restructured the code (with the help of ruediste) to migrate to a maven build structure.
Therefore if you check out the code now, you are not required to handle the Eclipse stuff anymore, because it has been moved to a separate project (maven downloads and builds the stuff for you now)

You're right that ideally the batchmode should not need to do any Swing stuff, but this is really an architectural issue which is not easy to solve without major refactorings. Although the batchmode does not instantiate the BaseGUI (and the standalone specific StandaloneGUI class), it does init some Swing related stuff in the DrawHandler/DrawPanel.

Perhaps you want to give it another try and check out the UMLet project. You should be able to ignore the eclipse plugin and gwt specific modules now and focus on umlet-swing and umlet-standalone.

By the way
> Unfortunately it's not always easy to find the correct way to do things in Swing ;)
was meant as a comment about the Swing classes which are used in normal GUI-mode (not batchmode)
After reading some examples and stackoverflow articles it seems OK for Swing classes to terminate using System.exit(0) and it's hard to know when to do stuff with Swing.invokeLater, ...
All of this should not be used in batchmode anyways (as far as I know)

**Comment by JensErat** [comment]
Not having the Eclipse dependencies any more is great news, and I will gladly put some effort in this issue as soon as possible. I'll very likely be using UMLet for some diagrams in my thesis and will have to fix this issue before doing so. ;)

**Comment by afdia** [comment]
Thanks! :)
By the way check out the dev info wiki page, I've updated it for the maven build: https://github.com/umlet/umlet/wiki/Information-for-Developers
If you have further suggestions please let me know

**Comment by JensErat** [comment]
The issue still exists with UMLet 14.4.1.

On my first try I failed at setting it up in Eclipse because of missing classes, and `maven clean install` failed because some result PNG differs from what it should be.

I'll have another look at it in a few weeks, currently I'm too busy.

**Comment by afdia** [comment]
Hi

Some classes are missing on purpose (because derived resources should not be checked in), but as they are generated during the first "mvn clean install" run, this problem should only occur once after the initial checkout.

The png comparison failure is interesting. I've added a testcase which makes sure that export functions work. The test works on my local Win7 PC and on the Travis CI.

Of course it would be nice if the test would work on every system (perhaps you can post your OS and java implementation and version)
In the meantime you can skip the test execution during a maven build with `mvn clean install -DskipTests`. Then the build should work.

**Comment by jGleitz** [comment]
Hi everybody!

On my machine (Ubuntu 15.10), when running a command like 

```
java -jar /opt/umlet/umlet.jar -action=convert -format=pdf -filename="file.uxf"
```

UMLet

 * starts
 * prints `Conversion finished: "path/file.uxf" to "null"`
 * actually converted the file
 * but does not return after the message has been printed.


However, this happens only on **OpenJDK**. Using **Oracle Java**, the conversion takes place and returns as expected.

```
$ /usr/lib/jvm/java-8-oracle/jre/bin/java -version
java version "1.8.0_72"
Java(TM) SE Runtime Environment (build 1.8.0_72-b15)
Java HotSpot(TM) 64-Bit Server VM (build 25.72-b15, mixed mode)

$ /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java -version
openjdk version "1.8.0_66-internal"
OpenJDK Runtime Environment (build 1.8.0_66-internal-b17)
OpenJDK 64-Bit Server VM (build 25.66-b17, mixed mode)

$ uname --all
$ uname -a
Linux Knuth 4.2.0-30-generic #36-Ubuntu SMP Fri Feb 26 00:58:07 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

**Comment by afdia** [comment]
Thanks, perhaps it's really linked to the Java version. All previous comments also use OpenJDK for Linux.
If I find some time I will setup a VM and check it on my local machine.

**Comment by afdia** [comment]
I've created a VirtualBox image with latest Linux Mint 17.3 "Rosa" - Cinnamon (64-bit) version and unfortunately I cannot reproduce the issue with the following OpenJDK version:
```
openjdk version "1.8.0_72-internal"
OpenJDK Runtime Environment (build 1.8.0_72-internal-b15)
OpenJDK 64-Bit Server VM (build 25.72-b15, mixed mode)
```

I followed the instructions on http://ubuntuhandbook.org/index.php/2015/01/install-openjdk-8-ubuntu-14-04-12-04-lts/ to install openjdk8

Afterwards I downloaded Umlet 14.1.1, unzipped it, switched into the directory and used
```
java -jar umlet.jar -action=convert -format=pdf -filename="palettes/UML Class.uxf"
```
which does not freeze. The pdf is created as expected in "palettes/UML Class.uxf.pdf"
The output was:
```
log4j:WARN No appenders could be found for logger (com.baselet.control.enums.Program).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
Conversion finished: "/home/afdia/Downloads/Umlet14_1_1/palettes/UML Class.uxf" to "/home/afdia/Downloads/Umlet14_1_1/palettes/UML Class.uxf.pdf"
```

Perhaps there is another factor which is causing the freezing, but I would suggest to try this exact command to make sure it's not related to the content of the uxf file or the exact line how the conversion is called.

**Comment by JensErat** [comment]
```
| <Fr 4. Mär 13:28:49 CET 2016> jenserat@t440s:/tmp/Umlet	(:0)
|  130 [508]$ java -version
java version "1.7.0_91"
OpenJDK Runtime Environment (IcedTea 2.6.3) (7u91-2.6.3-1)
OpenJDK 64-Bit Server VM (build 24.91-b01, mixed mode)
| <Fr 4. Mär 13:28:51 CET 2016> jenserat@t440s:/tmp/Umlet	(:0)
|    0 [509]$ java -jar umlet.jar -action=convert -format=pdf -filename="palettes/UML Class.uxf"
log4j:WARN No appenders could be found for logger (com.baselet.control.enums.Program).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
Conversion finished: "/tmp/Umlet/palettes/UML Class.uxf" to "/tmp/Umlet/palettes/UML Class.uxf.pdf
^C| <Fr 4. Mär 13:30:43 CET 2016> jenserat@t440s:/tmp/Umlet	(:0)
|  130 [510]$ uname -a
Linux t440s 4.3.0-1-amd64 #1 SMP Debian 4.3.5-1 (2016-02-06) x86_64 GNU/Linux
```

Execution was stuck for about two minutes until I interrupted. The PDF file exists, but gets deleted as soon as execution is interrupted by `^C`.

I will try to put together a virtual machine which reproduces the issue as soon as I finished some other more important tasks, trying not only to provide a system allowing to reproduce the issue, but also possibly finding the exact issue here.

It seems to be something related to the desktop environment, if I drop to another shell without an X server attached it works fine. Digging deeper, I realizeed `unset DISPLAY` resolves the issue. `export`ing it to the old value `DISPLAY=:0` results in stuck conversion again.

I'm running i3wm as display manager, which probably only provides a rather barebone-implementation of some features, so UMlet might hit some corner cases here.

**Comment by afdia** [comment]
Good idea, a VM image could be used to get a consistent environment where possible fixes could be tested.

**Comment by JensErat** [comment]
`apt-get install i3` and logging into this window manager might already be sufficient.

**Comment by afdia** [comment]
I used `apt-get install i3` and `apt-get install i3status` (otherwise i3 wouldnt work)

Afterwards I logged in, opened a terminal and used the same command.
Everthing worked as expected.

Perhaps this is related to specific hardware or drivers? I get the feeling that this isnt something which we would fix in our code anyway.

**Comment by olberger** [comment]
I had the very same issue and can confirm that the following helps:
`DISPLAY= umlet -action=convert -format=png -filename=diagram.uxf -output=diagram.png`
